roadmap/implementers-guide/src/node/disputes/dispute-distribution.md

```diff
@@ -1,3 +1,351 @@
+issue as in other subsystems. Rate limiting should be implemented at the
+substrate level, see
+[#7750](https://github.com/paritytech/substrate/issues/7750). Still we should
+make sure that it is not possible via spamming to prevent a dispute concluding
+or worse from getting noticed.
+
+Considered attack vectors:
+
+1. Invalid disputes (candidate does not exist) could make us
+   run out of resources. E.g. if we recorded every statement, we could run out
+   of disk space eventually.
+2. An attacker can just flood us with notifications on any notification
+   protocol, assuming flood protection is not effective enough, our unbounded
+   buffers can fill up and we will run out of memory eventually.
+3. Attackers could spam us at a high rate with invalid disputes. Our incoming
+   queue of requests could get monopolized by those malicious requests and we
+   won't be able to import any valid disputes and we could run out of resources,
+   if we tried to process them all in parallel.
+
+For tackling 1, we make sure to not occupy resources before we don't know a
+candidate is available. So we will not record statements to disk until we
+recovered availability for the candidate or know by some other means that the
+dispute is legit.
+
+For 2, we will pick up on any dispute on restart, so assuming that any realistic
+memory filling attack will take some time, we should be able to participate in a
+dispute under such attacks.
+
+For 3, full monopolization of the incoming queue should not be possible assuming
+substrate handles incoming requests in a somewhat fair way. Still we want some
+defense mechanisms, at the very least we need to make sure to not exhaust
+resources.
+
+The dispute coordinator will notify us
+via `DisputeDistributionMessage::ReportCandidateUnavailable` about unavailable
+candidates and we can disconnect from such peers/decrease their reputation
+drastically. This alone should get us quite far with regards to queue
+monopolization, as availability recovery is expected to fail relatively quickly
+for unavailable data.
+
+Still if those spam messages come at a very high rate, we might still run out of
+resources if we immediately call `DisputeCoordinatorMessage::ImportStatements`
+on each one of them. Secondly with our assumption of 1/3 dishonest validators,
+getting rid of all of them will take some time, depending on reputation timeouts
+some of them might even be able to reconnect eventually.
+
+To mitigate those issues we will process dispute messages with a maximum
+parallelism `N`. We initiate import processes for up to `N` candidates in
+parallel. Once we reached `N` parallel requests we will start back pressuring on
+the incoming requests. This saves us from resource exhaustion.
+
+To reduce impact of malicious nodes further, we can keep track from which nodes the
+currently importing statements came from and will drop requests from nodes that
+already have imports in flight.
```

Comment:
Ok, so the idea is that if a node tries to send K requests, that we'd begin processing 1 and then drop the K-1 other requests. An honest node is expected to retry, and in the meantime, we'll not accept requests from the other node. As we detect the data is not available, we drop the dispute and punish the bad peer. Does that sound accurate?

Reply:
Almost. We receive a request from node A - we will start importing votes. We will now drop any requests from node A until the import finished. This even works for distributed attacks - we will drop requests of up to `N` nodes and after we discovered it was spam, we will disconnect - freeing up resources even more. Assuming substrate networking is sane (and it will not let a single source monopolize the incoming channel) this is also already a second level defense. So we really should be fine here. The most important thing (apart from preventing monopolization - which can only be handled effectively at the substrate level) is, that we won't participate and thus escalate in a spam dispute, as this would amplify the attackers resources by the power of all nodes.